index page implementation

acrilique opened the pull request.

acrilique committed:
refactor(pdb): prepare for index page implementation

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,29 +234,79 @@ impl PageFlags {
+    Index,
+    /// The page is of an unknown or unsupported format.
+    Unknown,
+}
```
consider adding an `impl` with `get_data() -> Option<DataPageContent>`. That should reduce some of the more verbose pattern matching in the API.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -324,6 +374,18 @@ pub struct Page {
     /// Used space in bytes in the data section of the page.
     pub used_size: u16,
+    /// The content of the page.
+    #[br(parse_with = read_page_content)]
+    #[br(args(page_flags, page_start_pos, page_size, num_rows_small, page_type))]
+    #[bw(write_with = write_page_content, args(page_size))]
+    pub content: PageContent,
```
This is surprising to me. Are you sure the fields before the content here are still the same for index pages? I haven't looked into your research too far yet. Or to reword: Are you sure this refactoring is sufficient to account for the changes needed for the index pages?

acrilique replied:
About this, well, I wouldn't say it's written in stone. But this is sufficient to incorporate the structure of index pages as interpreted by @kimtore in the `rex` repo.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -359,15 +421,14 @@ pub struct Page {
     ///
     /// **Note:** This is a virtual field and not actually read from the file.
-    #[br(temp, calc = page_start_pos + u64::from(Self::HEADER_SIZE))]
+    #[br(temp, calc = page_start_pos + u64::from(Page::HEADER_SIZE))]
```
since we now split the page header, we should reflect that in the `Page::HEADER_SIZE`

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,29 +234,79 @@ impl PageFlags {
+    args: (PageFlags, u64, u32, u8, PageType),
+) -> BinResult<PageContent> {
+    let (page_flags, page_start_pos, page_size, num_rows_small, page_type) = args;
+    if page_flags.is_index_page() {
```
you can replace this using idiomatic `binrw`. Use `#[br(pre_assert(page_flags.is_index_page()))]` on each index variant.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,29 +234,79 @@ impl PageFlags {
+                page_size,
+                num_rows_small,
+                page_type,
+            },
```
instead of unpacking above and then repacking, just take the args for the `DataPageContent` in a single struct/tuple.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,29 +234,79 @@ impl PageFlags {
     args: (u32,),
 ) -> BinResult<()> {
-    let (page_size,) = args;
+    if let PageContent::Data(data_content) = page_content {
```
implement on `DataPageContent` directly.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -347,7 +409,7 @@ pub struct Page {
     ///
     /// **Note:** This is a virtual field and not actually read from the file.
-    #[br(temp, calc = Self::calculate_num_rows(num_rows_small, num_rows_large))]
+    #[br(temp, calc = Page::calculate_num_rows(num_rows_small, num_rows_large))]
```
also consider moving `calculate_num_rows` from `Page` to `DataPageContent`

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -359,15 +421,14 @@ pub struct Page {
-    #[br(if(num_row_groups > 0 && page_flags.page_has_data()))]
-    #[bw(write_with = write_page_contents, args(page_size))]
+    #[bw(ignore)]
     pub row_groups: Vec<RowGroup>,
```
I assume `DataPageContent` should only be written through write_page_content? In that case its probably most idiomatic to write a `BinWrite` impl yourself. Its a little boilerplate but its fine and the cleaner solution. using `write_with` only makes sense when you write part of an type declaratively and some part procedurally, but now that you split the type, you write the entire new type (`DataPageContent`) procedurally, just add a `BinWrite` impl (and remove the type-level `#[brw(...)]` on line 387 with `#[bw(...)]`).

Swiftb0y reviewed (changes requested):
thank you, this looks fine on a high-level, but there is still some potential for improvement.

acrilique committed:
refactor(pdb): address feedback from @Swiftb0y

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,31 +234,43 @@ impl PageFlags {
+    #[br(pre_assert(page_flags.is_index_page()))]
+    Index,
+    /// The page is of an unknown or unsupported format.
+    #[br(pre_assert(!page_flags.page_has_data() && !page_flags.is_index_page()))]
```
NIT, you can remove this IMO. It is implied.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -359,27 +382,73 @@ pub struct Page {
     #[br(map(|mut vec: Vec<RowGroup>| {vec.reverse(); vec}))]
-    #[br(if(num_row_groups > 0 && page_flags.page_has_data()))]
-    #[bw(write_with = write_page_contents, args(page_size))]
+    #[bw(ignore)]
```
now that you no longer derive binwrite, you can remove the bw directives. I'm confused why they don't break the build anyway.

Swiftb0y reviewed (commented):
two more nits. Rest looks pretty good. We don't need to polish this anyways because I'm sure that the structure probably needs to change again anyways during the course of further research.

acrilique committed:
refactor(pdb): address NITs from @Swiftb0y

acrilique committed:
feat(pdb): initial index page implementation

acrilique commented on src/pdb/mod.rs:
```diff
@@ -234,31 +235,131 @@ impl PageFlags {
+                .finish()
+        }
+    }
+}
```
Not sure if this is how I should have done it but it looks good in the output

Swiftb0y replied:
yeah looks good for now.

Swiftb0y replied:
you may also look into the modular-bitfield crate we already use for ANLZ files.

acrilique replied:
have just tried using that crate but I ended up with more verbose/complicated code. I'll just leave it as is for now

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,31 +235,131 @@ impl PageFlags {
+    pub unknown_a: u16,
+    /// Unknown field, usually `1fff` or `0000`.
+    pub unknown_b: u16,
+    /// Magic value, must be `0x03ec`.
+    #[br(temp, assert(magic1 == 0x03ec))]
+    #[bw(calc = 0x03ec)]
+    magic1: u16,
```
any reason you chose this formulation instead of just

```suggestion
    #[brw(magic(0x03ec))]
```

acrilique replied:
Well mainly because it is used in the magic value in `pub struct Page` and I was not aware of the existence of that formulation you just suggested. Shall I change that one too?

Swiftb0y replied:
yes please.

acrilique replied:
Done, also replaced in 2 magic fields found in the `Header` struct. Not 100% sure if this is desired, I can amend if necessary

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,31 +235,131 @@ impl PageFlags {
+    /// Must be zero.
+    #[br(temp, assert(zeros == 0))]
+    #[bw(calc = 0)]
+    zeros: u32,
```
same here

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,31 +235,131 @@ impl PageFlags {
+    /// The page is an index page.
+    #[br(pre_assert(page_flags.is_index_page()))]
+    Index(IndexPageContent),
+    /// The page is of an unknown or unsupported format.
+    Unknown,
```
this is where we might want to implement that previous suggestion of printing the bytes of the unknown page content as a hexdump.

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -393,14 +547,11 @@ impl Page {
+                DataPageContent::calculate_num_rows(self.num_rows_small, content.num_rows_large)
+            }
+            PageContent::Index(_) | PageContent::Unknown => 0,
         }
```
consider moving to DataPageContent, since the values chosen for the other alternatives are somewhat meaningless, let the caller unwrap the enum if they want to know the row count. They likely know better how to handle the pagecontent types.

acrilique replied:
just noticed that method and `has_data` are now unused, so I removed them. DataPageContent already has a `num_rows` temp field which could be made public. also `has_data` makes no sense given that now the workflow consists on calling `content.into_data()` and checking the result

Swiftb0y replied:
Yeah that makes sense.

Swiftb0y commented on src/main.rs:
```diff
@@ -154,12 +155,20 @@ fn dump_pdb(path: &PathBuf, typ: DatabaseType) -> rekordcrate::Result<()> {
+                    println!("      {:?}", entry);
                 }
-            })
+            }
```
if you actually do explicitly pattern match here, you don't need the extra clone (instead of using `get_*` function).

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -234,31 +235,131 @@ impl PageFlags {
+            PageContent::Index(index) => Some(index),
+            _ => None,
+        }
     }
```
Nit: rename these as `into_*`. `get_` usually works on `&self`, `get_mut_*` on `&mut self` and `into_*` consumes `self`. Its very common to have these three variants present (there is no way around to implementing all three if you need them all afaik). You shouldn't implement all three here, just saying consider changing the name so it aligns with the conventions from `std`.

acrilique committed:
refactor(pdb): rename get_* methods to into_* to align with `std`

acrilique committed:
refactor(pdb): `#[br(temp, assert/#[bw(calc` -> `#[brw(magic` in structs

acrilique committed:
fix(pdb): explicit pattern matching to avoid extra `clone()`

acrilique committed:
refactor(pdb): remove unused methods from `Page` impl

Swiftb0y commented:
this looks good. do you want to go ahead and open a PR to the upstream repo?

acrilique committed:
docs(pdb): enhance documentation for some index page fields

acrilique committed:
feat(pdb): add methods to create new and empty `IndexEntry` instances

acrilique committed:
fix(pdb): custom `BinWrite` for `IndexPageContent` to comply with binary

acrilique committed:
test(pdb): add `index_page` unit test

acrilique commented:
Just added a passing unit test (had to add `IndexEntry` constructors and implement a custom `BinWrite`). 
I will now open a PR to upstream

Swiftb0y commented on src/pdb/mod.rs:
```diff
@@ -262,6 +264,20 @@ impl IndexEntry {
+        assert!(page_index.0 <= 0x1FFF_FFFF);
+        assert!(index_flags & 0b111 == index_flags);
+        Self((page_index.0 << 3) | u32::from(index_flags & 0b111))
+    }
```
return an `Option<Self>`, let the caller call `.unwrap()` if they deem panics to be okay. move the `PageIndex` invariant to PageIndex itself (also return an `Option<Self>` there). Consider making this and below `const fn`.

acrilique committed:
chore(pdb): add reference to `binrw` issue for custom padding value

acrilique committed:
fix(pdb): address feedback by @Swiftb0y

- make `IndexEntry` constructor return `Option`.
- add `PageIndex` `new()`. returns `Option` and checks max value isn't surpassed
- make `new` and `empty` in `IndexEntry` consts

acrilique committed:
refactor(pdb): address NITs from @Holzhaus and make clippy happy

- changed `new()` to `try_from()` in `PageIndex` and `IndexEntry`
- added PdbError enum
- use `u64::from()` instead of `as u64`
- change multi-line comments for single-line, nestable comments

acrilique committed:
fix(pdb): only allow PageIndex values less than 0x03FF_FFFF
